File: Software/StarPRNT SDK/ObjectiveC SDK/ObjectiveC SDK/Communication.m
//
//  Communication.m
//  ObjectiveC SDK
//
//  Created by Yuji on 2015/**/**.
//  Copyright (c) 2015年 Star Micronics. All rights reserved.
//

#import <UIKit/UIKit.h>

#import "Communication.h"

@implementation Communication

+ (BOOL)sendCommands:(NSData *)commands port:(SMPort *)port {
    BOOL result = NO;
    
    NSString *title   = @"";
    NSString *message = @"";
    
    uint32_t commandLength = (uint32_t) commands.length;
    
    unsigned char *commandsBytes = (unsigned char *) commands.bytes;
    
    @try {
        while (YES) {
            if (port == nil) {
                title = @"Fail to Open Port";
                break;
            }
            
            StarPrinterStatus_2 printerStatus;
            
            [port beginCheckedBlock:&printerStatus :2];
            
            if (printerStatus.offline == SM_TRUE) {
                title   = @"Printer Error";
                message = @"Printer is offline (BeginCheckedBlock)";
                break;
            }
            
            NSDate *startDate = [NSDate date];
            
            uint32_t total = 0;
            
            while (total < commandLength) {
                uint32_t written = [port writePort:commandsBytes :total :commandLength - total];
                
                total += written;
                
                if ([[NSDate date] timeIntervalSinceDate:startDate] >= 30.0) {     // 30000mS!!!
                    break;
                }
            }
            
            if (total < commandLength) {
                title   = @"Printer Error";
                message = @"Write port timed out";
                break;
            }
            
            port.endCheckedBlockTimeoutMillis = 30000;     // 30000mS!!!
            
            [port endCheckedBlock:&printerStatus :2];
            
            if (printerStatus.offline == SM_TRUE) {
                title   = @"Printer Error";
                message = @"Printer is offline (endCheckedBlock)";
                break;
            }
            
            title   = @"Send Commands";
            message = @"Success";
            
            result = YES;
            break;
        }
    }
    @catch (PortException *exc) {
        title   = @"Printer Error";
        message = @"Write port timed out (PortException)";
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:title message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        
        [alertView show];
    });
    
    return result;
}

+ (BOOL)sendCommandsDoNotCheckCondition:(NSData *)commands port:(SMPort *)port {
    BOOL result = NO;
    
    NSString *title   = @"";
    NSString *message = @"";
    
    uint32_t commandLength = (uint32_t) commands.length;
    
    unsigned char *commandsBytes = (unsigned char *) commands.bytes;
    
    @try {
        while (YES) {
            if (port == nil) {
                title = @"Fail to Open Port";
                break;
            }
            
            StarPrinterStatus_2 printerStatus;
            
            [port getParsedStatus:&printerStatus :2];
            
//          if (printerStatus.offline == SM_TRUE) {     // Do not check condition.
//              title   = @"Printer Error";
//              message = @"Printer is offline (BeginCheckedBlock)";
//              break;
//          }
            
            NSDate *startDate = [NSDate date];
            
            uint32_t total = 0;
            
            while (total < commandLength) {
                uint32_t written = [port writePort:commandsBytes :total :commandLength - total];
                
                total += written;
                
                if ([[NSDate date] timeIntervalSinceDate:startDate] >= 30.0) {     // 30000mS!!!
                    break;
                }
            }
            
            if (total < commandLength) {
                title   = @"Printer Error";
                message = @"Write port timed out";
                break;
            }
            
            [port getParsedStatus:&printerStatus :2];
            
//          if (printerStatus.offline == SM_TRUE) {     // Do not check condition.
//              title   = @"Printer Error";
//              message = @"Printer is offline (endCheckedBlock)";
//              break;
//          }
            
            title   = @"Send Commands";
            message = @"Success";
            
            result = YES;
            break;
        }
    }
    @catch (PortException *exc) {
        title   = @"Printer Error";
        message = @"Write port timed out (PortException)";
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:title message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        
        [alertView show];
    });
    
    return result;
}

+ (BOOL)sendCommands:(NSData *)commands portName:(NSString *)portName portSettings:(NSString *)portSettings timeout:(NSInteger)timeout {
    BOOL result = NO;
    
    NSString *title   = @"";
    NSString *message = @"";
    
    if (timeout > UINT32_MAX) {
        timeout = UINT32_MAX;
    }
    
    uint32_t commandLength = (uint32_t) commands.length;
    
    unsigned char *commandsBytes = (unsigned char *) commands.bytes;
    
    SMPort *port = nil;
    
    @try {
        while (YES) {
            port = [SMPort getPort:portName :portSettings :(uint32_t) timeout];
            
            if (port == nil) {
                title = @"Fail to Open Port";
                break;
            }
            
            StarPrinterStatus_2 printerStatus;
            
            [port beginCheckedBlock:&printerStatus :2];
            
            if (printerStatus.offline == SM_TRUE) {
                title   = @"Printer Error";
                message = @"Printer is offline (BeginCheckedBlock)";
                break;
            }
            
            NSDate *startDate = [NSDate date];
            
            uint32_t total = 0;
            
            while (total < commandLength) {
                uint32_t written = [port writePort:commandsBytes :total :commandLength - total];
                
                total += written;
                
                if ([[NSDate date] timeIntervalSinceDate:startDate] >= 30.0) {     // 30000mS!!!
                    break;
                }
            }
            
            if (total < commandLength) {
                title   = @"Printer Error";
                message = @"Write port timed out";
                break;
            }
            
            port.endCheckedBlockTimeoutMillis = 30000;     // 30000mS!!!
            
            [port endCheckedBlock:&printerStatus :2];
            
            if (printerStatus.offline == SM_TRUE) {
                title   = @"Printer Error";
                message = @"Printer is offline (endCheckedBlock)";
                break;
            }
            
            title   = @"Send Commands";
            message = @"Success";
            
            result = YES;
            break;
        }
    }
    @catch (PortException *exc) {
        title   = @"Printer Error";
        message = @"Write port timed out (PortException)";
    }
    @finally {
        if (port != nil) {
            [SMPort releasePort:port];
        }
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:title message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        
        [alertView show];
    });
    
    return result;
}

+ (BOOL)sendCommandsDoNotCheckCondition:(NSData *)commands portName:(NSString *)portName portSettings:(NSString *)portSettings timeout:(NSInteger)timeout {
    BOOL result = NO;
    
    NSString *title   = @"";
    NSString *message = @"";
    
    if (timeout > UINT32_MAX) {
        timeout = UINT32_MAX;
    }
    
    uint32_t commandLength = (uint32_t) commands.length;
    
    unsigned char *commandsBytes = (unsigned char *) commands.bytes;
    
    SMPort *port = nil;
    
    @try {
        while (YES) {
            port = [SMPort getPort:portName :portSettings :(uint32_t) timeout];
            
            if (port == nil) {
                title = @"Fail to Open Port";
                break;
            }
            
            StarPrinterStatus_2 printerStatus;
            
            [port getParsedStatus:&printerStatus :2];
            
//          if (printerStatus.offline == SM_TRUE) {     // Do not check condition.
//              title   = @"Printer Error";
//              message = @"Printer is offline (BeginCheckedBlock)";
//              break;
//          }
            
            NSDate *startDate = [NSDate date];
            
            uint32_t total = 0;
            
            while (total < commandLength) {
                uint32_t written = [port writePort:commandsBytes :total :commandLength - total];
                
                total += written;
                
                if ([[NSDate date] timeIntervalSinceDate:startDate] >= 30.0) {     // 30000mS!!!
                    break;
                }
            }
            
            if (total < commandLength) {
                title   = @"Printer Error";
                message = @"Write port timed out";
                break;
            }
            
            [port getParsedStatus:&printerStatus :2];
            
//          if (printerStatus.offline == SM_TRUE) {     // Do not check condition.
//              title   = @"Printer Error";
//              message = @"Printer is offline (endCheckedBlock)";
//              break;
//          }
            
            title   = @"Send Commands";
            message = @"Success";
            
            result = YES;
            break;
        }
    }
    @catch (PortException *exc) {
        title   = @"Printer Error";
        message = @"Write port timed out (PortException)";
    }
    @finally {
        if (port != nil) {
            [SMPort releasePort:port];
        }
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:title message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        
        [alertView show];
    });
    
    return result;
}

+ (void)connectBluetooth {
    [[EAAccessoryManager sharedAccessoryManager] showBluetoothAccessoryPickerWithNameFilter:nil completion:^(NSError *error) {
        BOOL result;
        BOOL show;
        
        if (error != nil) {
            NSLog(@"Error:%@", error.description);
            
            switch (error.code) {
                case EABluetoothAccessoryPickerAlreadyConnected :
                    result = YES;
                    show   = YES;
                    break;
                case EABluetoothAccessoryPickerResultCancelled :
                case EABluetoothAccessoryPickerResultFailed    :
                    result = NO;
                    show   = NO;
                    break;
                default                                       :
//              case EABluetoothAccessoryPickerResultNotFound :
                    result = NO;
                    show   = YES;
                    break;
            }
        }
        else {
            result = YES;
            show   = YES;
        }
        
        if (show == YES) {
            dispatch_async(dispatch_get_main_queue(), ^{
                UIAlertView *alertView;
                
                if (result == YES) {
                    alertView = [[UIAlertView alloc] initWithTitle:@"Success" message:nil delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
                }
                else {
                    alertView = [[UIAlertView alloc] initWithTitle:@"Fail to Connect" message:nil delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
                }
                
                [alertView show];
            });
        }
    }];
}

+ (BOOL)disconnectBluetooth:(NSString *)modelName portName:(NSString *)portName portSettings:(NSString *)portSettings timeout:(NSInteger)timeout {
    BOOL result = NO;
    
    SMPort *port = nil;
    
    @try {
        while (YES) {
            port = [SMPort getPort:portName :portSettings :(uint32_t) timeout];
            
            if (port == nil) {
                break;
            }
            
            if ([modelName hasPrefix:@"TSP143IIIBI"]) {
                unsigned char commandBytes[] = {0x1b, 0x1c, 0x26, 0x49};     // Only TSP143IIIBI
                
                uint32_t commandLength = sizeof(commandBytes);
                
                StarPrinterStatus_2 printerStatus;
                
                [port beginCheckedBlock:&printerStatus :2];
                
                if (printerStatus.offline == SM_TRUE) {
                    break;
                }
                
                NSDate *startDate = [NSDate date];
                
                uint32_t total = 0;
                
                while (total < commandLength) {
                    uint32_t written = [port writePort:commandBytes :total :commandLength - total];
                    
                    total += written;
                    
                    if ([[NSDate date] timeIntervalSinceDate:startDate] >= 30.0) {     // 30000mS!!!
                        break;
                    }
                }
                
                if (total < commandLength) {
                    break;
                }
                
//              port.endCheckedBlockTimeoutMillis = 30000;     // 30000mS!!!
//
//              [port endCheckedBlock:&printerStatus :2];
//
//              if (printerStatus.offline == SM_TRUE) {
//                  break;
//              }
            }
            else {
                if ([port disconnect] == NO) {
                    break;
                }
            }
        
            result = YES;
            break;
        }
    }
    @catch (PortException *exception) {
    }
    @finally {
        if (port != nil) {
            [SMPort releasePort:port];
        }
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alertView;
        
        if (result == YES) {
            alertView = [[UIAlertView alloc] initWithTitle:@"Success" message:nil delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        }
        else {
            alertView = [[UIAlertView alloc] initWithTitle:@"Fail to Disconnect" message:@"Note. Portable Printers is not supported." delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        }
        
        [alertView show];
    });
    
    return result;
}

@end

File: Software/StarPRNT SDK/Swift SDK/Swift SDK/Communication.swift
//
//  Communication.swift
//  Swift SDK
//
//  Created by Yuji on 2015/**/**.
//  Copyright © 2015年 Star Micronics. All rights reserved.
//

import Foundation

class Communication {
    static func sendCommands(commands: NSData!, port: SMPort!) -> Bool {
        var result: Bool = false
        
        var title:   String = ""
        var message: String = ""
        
        var error: NSError?
        
        let length: UInt32 = UInt32(commands.length)
        
        var array: [UInt8] = [UInt8](count: commands.length, repeatedValue: 0)
        
        commands.getBytes(&array, length: commands.length)
        
        while true {
            if port == nil {
                title   = "Fail to Open Port"
                message = ""
                break
            }
            
            var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
            
            port.beginCheckedBlock(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
            if printerStatus.offline == sm_true {
                title   = "Printer Error"
                message = "Printer is offline (BeginCheckedBlock)"
                break
            }
            
            let startDate: NSDate = NSDate()
            
            var total: UInt32 = 0
            
            while total < length {
                let written: UInt32 = port.writePort(array, total, length - total, &error)
                
                if error != nil {
                    break
                }
                
                total += written
                
                if NSDate().timeIntervalSinceDate(startDate) >= 30.0 {     // 30000mS!!!
                    title   = "Printer Error"
                    message = "Write port timed out"
                    break
                }
            }
            
            if total < length {
                break
            }
            
            port.endCheckedBlockTimeoutMillis = 30000     // 30000mS!!!
            
            port.endCheckedBlock(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
            if printerStatus.offline == sm_true {
                title   = "Printer Error"
                message = "Printer is offline (endCheckedBlock)"
                break
            }
            
            title   = "Send Commands"
            message = "Success"
            
            result = true
            break
        }
        
        if error != nil {
            title   = "Printer Error"
            message = error!.description
        }
        
        dispatch_async(dispatch_get_main_queue()) {
            let alertView: UIAlertView = UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: "OK")
            
            alertView.show()
        }
        
        return result
    }
    
    static func sendCommandsDoNotCheckCondition(commands: NSData!, port: SMPort!) -> Bool {
        var result: Bool = false
        
        var title:   String = ""
        var message: String = ""
        
        var error: NSError?
        
        let length: UInt32 = UInt32(commands.length)
        
        var array: [UInt8] = [UInt8](count: commands.length, repeatedValue: 0)
        
        commands.getBytes(&array, length: commands.length)
        
        while true {
            if port == nil {
                title   = "Fail to Open Port"
                message = ""
                break
            }
            
            var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
            
            port.getParsedStatus(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
//          if printerStatus.offline == sm_true {     // Do not check condition.
//              title   = "Printer Error"
//              message = "Printer is offline (getParsedStatus)"
//              break
//          }
            
            let startDate: NSDate = NSDate()
            
            var total: UInt32 = 0
            
            while total < length {
                let written: UInt32 = port.writePort(array, total, length - total, &error)
                
                if error != nil {
                    break
                }
                
                total += written
                
                if NSDate().timeIntervalSinceDate(startDate) >= 30.0 {     // 30000mS!!!
                    title   = "Printer Error"
                    message = "Write port timed out"
                    break
                }
            }
            
            if total < length {
                break
            }
            
            port.getParsedStatus(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
//          if printerStatus.offline == sm_true {     // Do not check condition.
//              title   = "Printer Error"
//              message = "Printer is offline (getParsedStatus)"
//              break
//          }
            
            title   = "Send Commands"
            message = "Success"
            
            result = true
            break
        }
        
        if error != nil {
            title   = "Printer Error"
            message = error!.description
        }
        
        dispatch_async(dispatch_get_main_queue()) {
            let alertView: UIAlertView = UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: nil, otherButtonTitles: "OK")
            
            alertView.show()
        }
        
        return result
    }
    
    static func sendCommands(commands: NSData!, portName: String!, portSettings: String!, timeout: UInt32) -> Bool! {
        var result: Bool = false
        
        var title:   String = ""
        var message: String = ""
        
        var error: NSError?
        
        let length: UInt32 = UInt32(commands.length)
        
        var array: [UInt8] = [UInt8](count: commands.length, repeatedValue: 0)
        
        commands.getBytes(&array, length: commands.length)
        
        while true {
            guard let port: SMPort = SMPort.getPort(portName, portSettings, timeout) else {
                title   = "Fail to Open Port"
                message = ""
                break
            }
            
            defer {
                SMPort.releasePort(port)
            }
            
            var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
            
            port.beginCheckedBlock(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
            if printerStatus.offline == sm_true {
                title   = "Printer Error"
                message = "Printer is offline (BeginCheckedBlock)"
                break
            }
            
            let startDate: NSDate = NSDate()
            
            var total: UInt32 = 0
            
            while total < length {
                let written: UInt32 = port.writePort(array, total, length - total, &error)
                
                if error != nil {
                    break
                }
                
                total += written
                
                if NSDate().timeIntervalSinceDate(startDate) >= 30.0 {     // 30000mS!!!
                    title   = "Printer Error"
                    message = "Write port timed out"
                    break
                }
            }
            
            if total < length {
                break
            }
            
            port.endCheckedBlockTimeoutMillis = 30000     // 30000mS!!!
            
            port.endCheckedBlock(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
            if printerStatus.offline == sm_true {
                title   = "Printer Error"
                message = "Printer is offline (endCheckedBlock)"
                break
            }
            
            title   = "Send Commands"
            message = "Success"
            
            result = true
            break
        }
        
        if error != nil {
            title   = "Printer Error"
            message = error!.description
        }
        
        dispatch_async(dispatch_get_main_queue()) {
            let alertView: UIAlertView = UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: "OK")
            
            alertView.show()
        }
        
        return result
    }
    
    static func sendCommandsDoNotCheckCondition(commands: NSData!, portName: String!, portSettings: String!, timeout: UInt32) -> Bool! {
        var result: Bool = false
        
        var title:   String = ""
        var message: String = ""
        
        var error: NSError?
        
        let length: UInt32 = UInt32(commands.length)
        
        var array: [UInt8] = [UInt8](count: commands.length, repeatedValue: 0)
        
        commands.getBytes(&array, length: commands.length)
        
        while true {
            guard let port: SMPort = SMPort.getPort(portName, portSettings, timeout) else {
                title   = "Fail to Open Port"
                message = ""
                break
            }
            
            defer {
                SMPort.releasePort(port)
            }
            
            var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
            
            port.getParsedStatus(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
//          if printerStatus.offline == sm_true {     // Do not check condition.
//              title   = "Printer Error"
//              message = "Printer is offline (getParsedStatus)"
//              break
//          }
            
            let startDate: NSDate = NSDate()
            
            var total: UInt32 = 0
            
            while total < length {
                let written: UInt32 = port.writePort(array, total, length - total, &error)
                
                if error != nil {
                    break
                }
                
                total += written
                
                if NSDate().timeIntervalSinceDate(startDate) >= 30.0 {     // 30000mS!!!
                    title   = "Printer Error"
                    message = "Write port timed out"
                    break
                }
            }
            
            if total < length {
                break
            }
            
            port.getParsedStatus(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
//          if printerStatus.offline == sm_true {     // Do not check condition.
//              title   = "Printer Error"
//              message = "Printer is offline (getParsedStatus)"
//              break
//          }
            
            title   = "Send Commands"
            message = "Success"
            
            result = true
            break
        }
        
        if error != nil {
            title   = "Printer Error"
            message = error!.description
        }
        
        dispatch_async(dispatch_get_main_queue()) {
            let alertView: UIAlertView = UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: "OK")
            
            alertView.show()
        }
        
        return result
    }

    static func connectBluetooth() {
        EAAccessoryManager.sharedAccessoryManager().showBluetoothAccessoryPickerWithNameFilter(nil) { (error) -> Void in
            var result: Bool
            var show:   Bool
            
            if error != nil {
                NSLog("Error:%@", error!.description)
                
                switch error!.code {
                case EABluetoothAccessoryPickerErrorCode.AlreadyConnected.rawValue :
                    result = true
                    show   = true
                case EABluetoothAccessoryPickerErrorCode.ResultCancelled.rawValue,
                     EABluetoothAccessoryPickerErrorCode.ResultFailed   .rawValue :
                    result = false
                    show   = false
//              case EABluetoothAccessoryPickerErrorCode.ResultNotFound :
                default                                                 :
                    result = false
                    show   = true
                }
            }
            else {
                result = true
                show   = true
            }
            
            if show == true {
                dispatch_async(dispatch_get_main_queue()) {
                    var alertView: UIAlertView
                    
                    if result == true {
                        alertView = UIAlertView(title: "Success", message: nil, delegate: nil, cancelButtonTitle: "OK")
                    }
                    else {
                        alertView = UIAlertView(title: "Fail to Connect", message: nil, delegate: nil, cancelButtonTitle: "OK")
                    }
                    
                    alertView.show()
                }
            }
        }
    }
    
    static func disconnectBluetooth(modelName: String!, portName: String!, portSettings: String!, timeout: UInt32) -> Bool! {
        var result: Bool = false
        
        while true {
            guard let port: SMPort = SMPort.getPort(AppDelegate.getPortName(), AppDelegate.getPortSettings(), 10000) else {     // 10000mS!!!
                break
            }
            
            defer {
                SMPort.releasePort(port)
            }
            
            if modelName.hasPrefix("TSP143IIIBI") == true {
                var error: NSError?
                
                let array: [UInt8] = [0x1b, 0x1c, 0x26, 0x49]     // Only TSP143IIIBI
                
                let length: UInt32 = UInt32(array.count)
                
                var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
                
                port.beginCheckedBlock(&printerStatus, 2, &error)
                
                if error != nil {
                    break
                }
                
                if printerStatus.offline == sm_true {
                    break
                }
                
                let startDate: NSDate = NSDate()
                
                var total: UInt32 = 0
                
                while total < length {
                    let written: UInt32 = port.writePort(array, total, length - total, &error)
                    
                    if error != nil {
                        break
                    }
                    
                    total += written
                    
                    if NSDate().timeIntervalSinceDate(startDate) >= 30.0 {     // 30000mS!!!
                        break
                    }
                }
                
                if total < length {
                    break
                }
                
//              port.endCheckedBlockTimeoutMillis = 30000     // 30000mS!!!
//
//              port.endCheckedBlock(&printerStatus, 2, &error)
//
//              if error != nil {
//                  break
//              }
//
//              if printerStatus.offline == sm_true {
//                  break
//              }
            }
            else {
                if port.disconnect() == false {
                    break
                }
            }
            
            result = true
            break
        }
        
        dispatch_async(dispatch_get_main_queue()) {
            let alertView: UIAlertView
            
            if result == true {
                alertView = UIAlertView(title: "Success", message: nil, delegate: nil, cancelButtonTitle: "OK")
            }
            else {
                alertView = UIAlertView(title: "Fail to Disconnect", message: "Note. Portable Printers is not supported.", delegate: nil, cancelButtonTitle: "OK")
            }
            
            alertView.show()
        }
        
        return result
    }
}
